handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,471 @@
+            ssc.delete();
+        }
+    }
+
+    private static Channel newServer(EventLoopGroup group, SocketAddress address,
+            SslContext context, ChannelHandler handler) {
+
+        ServerBootstrap bootstrap = new ServerBootstrap()
+                .channel(LocalServerChannel.class)
+                .group(group)
+                .childHandler(newHandler(context, handler));
+
+        return bootstrap.bind(address)
+                .syncUninterruptibly()
+                .channel();
+    }
+
+    private static Channel newClient(EventLoopGroup group, SocketAddress address,
+            SslContext context, ChannelHandler handler) {
+
+        Bootstrap bootstrap = new Bootstrap()
+                .channel(LocalChannel.class)
+                .group(group)
+                .handler(newHandler(context, handler));
+
+        return bootstrap.connect(address)
+                .syncUninterruptibly()
+                .channel();
+    }
+
+    private static ChannelHandler newHandler(final SslContext context, final ChannelHandler handler) {
+        return new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                ChannelPipeline pipeline = ch.pipeline();
+
+                SSLEngine engine = context.newEngine(ch.alloc());
+                ospfySslEngine(engine);
+
+                pipeline.addLast(new SslHandler(engine));
+
+                if (handler != null) {
+                    pipeline.addLast(handler);
+                }
+            }
+        };
+    }
+
+    private static void ocspfySslContext(SslContext context, OcspContext ocspContext) {
+        ((ReferenceCountedOpenSslContext) context).ocspContext(ocspContext);
+    }
+
+    private static SSLEngine ospfySslEngine(SSLEngine engine) {
+        ((ReferenceCountedOpenSslEngine) engine).enableOcsp(true);
+        return engine;
+    }
+
+    private static ServerOcspContext newServerOcspContext() {
+        // Assume we got the OCSP staple from somewhere. Using a bogus byte[]
+        // in the test because getting a true staple from the CA is quite involved.
+        // It requires HttpCodec and Bouncycastle and the test may be very unreliable
+        // because the OCSP responder servers are basically being DDoS'd by the
+        // Internet.
+
+        byte[] bogus = "I am a bogus OCSP staple. OpenSSL does not care about the format of the byte[]!".getBytes();
+        return new DefaultServerOcspContext(bogus);
+    }
+
+    private static class TestClientOcspContext implements ClientOcspContext {
+
+        private final CountDownLatch latch = new CountDownLatch(1);
+
+        private final boolean valid;
+
+        private volatile byte[] staple;
+
+        public TestClientOcspContext(boolean valid) {
+            this.valid = valid;
+        }
+
+        public byte[] staple() throws InterruptedException, TimeoutException {
+            if (!latch.await(10L, TimeUnit.SECONDS)) {
```

Comment:
this is expected to be called in the junit thread right, if so can we just do `assertTrue`?